### PR TITLE
Connect Redirect Test fixes

### PR DESCRIPTION
**Pull request by matthewige**
## Description

https://github.com/microsoft/ebpf-for-windows/pull/2823 introduced a change which caused connect redirect tests to fail.

The following was updated:
- Connect (non-redirects) were incorrectly assuming redirection was occurring when it was not. This has been updated to check if redirection is expected or not.
- Querying the redirect context on TCP sockets was updated to now correctly query the accept socket (previously was listen socket).
- UDP tests did not have appropriate logic to query the redirect context. For now, UDP tests do not validate the redirect context was obtained. A future change will fix this: https://github.com/microsoft/ebpf-for-windows/issues/2853 
- Connect redirect tests were updated to explicitly bind to an IP address. This is necessary as these tests use duonic, which was previously treating this traffic as remote traffic instead of loopback.

## Testing

Updates connect redirect tests.

## Documentation

N/A

## Installation

N/A

### CODE REVIEW

**Review by dthaler — APPROVED** [review]

File: tests/connect_redirect/connect_redirect_tests.cpp [hunk 1]
```diff
@@ -379,13 +378,20 @@ connect_redirect_test(
...
-        // Remote redirection expects the SERVER_MESSAGE response.
-        // Local redirection expects the redirect context response.
+        // For local redirection, the redirect context is expected to be set and returned.
+        // If the connection is not redirected or is redirected to a remote address,
+        // check for the SERVER_MESSAGE generic response.
+        bool redirected =
+            (destination_port != proxy_port || !INETADDR_ISEQUAL((SOCKADDR*)&destination, (SOCKADDR*)&proxy));
+        bool local_redirect = !(
+            INETADDR_ISEQUAL((SOCKADDR*)&proxy, (SOCKADDR*)&_globals.addresses[socket_family_t::IPv4].remote_address) ||
+            INETADDR_ISEQUAL((SOCKADDR*)&proxy, (SOCKADDR*)&_globals.addresses[socket_family_t::IPv6].remote_address) ||
```
Comment (matthewige) [inline tests/connect_redirect/connect_redirect_tests.cpp:388]: Use addresses[_globals.family] instead.

File: tests/connect_redirect/connect_redirect_tests.cpp [hunk 2]
```diff
@@ -379,13 +378,20 @@ connect_redirect_test(
...
+            INETADDR_ISEQUAL((SOCKADDR*)&proxy, (SOCKADDR*)&_globals.addresses[socket_family_t::Dual].remote_address));
```
Comment (saxena-anurag) [inline tests/connect_redirect/connect_redirect_tests.cpp:389]: qq -- would this not return true for _all_ redirections? (both local and remote)?

Comment (matthewige) [reply tests/connect_redirect/connect_redirect_tests.cpp:389]: Can you double check my read here - 'proxy' will either be loopback, local, or remote address. If it is loopback or local, it should evaluate to !(false || false || false), which would be true. If the 'proxy' is the remote address, then it should evaluate to !(true || false || false) [in some order], which would be false.

Comment (saxena-anurag) [reply tests/connect_redirect/connect_redirect_tests.cpp:389]: You are right. I (wrongly) though the original destination is always saved in `remote_address`.

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**